Default Classifier General Improvement

Steven-Yiran opened the pull request:
Default Classifier now supports np arrays of shape > 2 like images.

huiwengoh commented:
Hi @Steven-Yiran, thanks for the contribution!

Adding support for images and more is great, I just have a few comments and a curious question:

- Your addition so far has made it possible to run `CleanLearning.find_label_issues()` with larger dimensional data, however I believe `CleanLearning.fit()` will fail due to this line over [here](https://github.com/cleanlab/cleanlab/blob/master/cleanlab/classification.py#L583) since the classifier is called again.
    - Similarly, `CleanLearning.predict()` and related functions would likely run into the same issues.

- Could you please add some unittests to ensure code coverage of the addition you made? It'll be super helpful to make sure the code works as expected! It would likely be placed in [tests/test_classification.py](https://github.com/cleanlab/cleanlab/blob/master/tests/test_classification.py) or [tests/test_filter_count.py](https://github.com/cleanlab/cleanlab/blob/master/tests/test_filter_count.py).

- And just out of curiosity, why only support larger dimensional data with the default (LogReg) classifier but not others?

Steven-Yiran commented:
Hi @huiwengoh, thanks a lot for your comments, they are really helpful.

I will try to fix the `cleanlearning.fit()` issues and add unit tests. 

For your last question, the original motivation for this pull request is the "Cleanlearning default classifier" issue (https://github.com/cleanlab/cleanlab/issues/406), so that is why I only focused on the default classifier. I guess the reason might be if users have provided a defined classifier then they will be responsible for making sure it works on the dataset they provided.

jwmueller commented:
I like @Steven-Yiran 's reasoning, the goal here is just to make CleanLearning work out of the box for more data types. If user has a specific classifier they want to use, it's their responsibility to ensure it works for the data (which I believe is actually already handled by the package for non 2D numpy arrays)

huiwengoh commented:
Ah yes that totally makes sense! I saw this PR outside the context of that issue so I was just wondering - thanks for the clarification!

Feel free to reach out if you have any other questions regarding making this work for the rest of the `CleanLearning` methods or the unittests 😄

Steven-Yiran commented:
Modifications:
1. Added high-dimension array conversion for all impacted methods in the CleanLearning class
2. Added tests for the default classifier and the dimension reduction function. 

Instead of writing the dimension reduction code in each method of the CleanLearning class, I added a class property in the CleanLearning class and a function in `utils.py` which is reused in the CleanLearing class.

I would love to get feedback on this design and the overall submission. Thanks!

jwmueller commented:
don't forget to run `black` formatter to ensure your code passes the style check

huiwengoh commented on cleanlab/classification.py:
```diff
@@ -602,8 +609,12 @@ def predict(self, *args, **kwargs) -> np.ndarray:
-
-        return self.clf.predict(*args, **kwargs)
+        if self.default_clf:
+            X = force_two_dimensions(args[0])
+            new_args = (X,) + args[1:]
```
@jwmueller What was the reasoning behind not specifying `X` as an argument explicitly? (given that that is what sklearn expects). Assuming the features are usually denoted by `X` or `x` we can check for that, but this will become non-trivial to support if users can pass in any arg name.

Steven-Yiran replied:
Thanks for catching the error! I think I had the same question while looking at the `predict()` and `predict_proba()` functions. Maybe the reason is if the classifier is defined by the user then it could expect an input dataset name different from `X`

huiwengoh reviewed (changes requested):
Good call adding that function in `utils`! The functionality is coming together, just had a few small comments - most important of which is to make sure `cl.predict(X=X)` works.

Also as mentioned above, don't forget to run the `black` style formatter for your code!

Steven-Yiran commented:
Added support for `kwargs` and modified related default classifier tests.

huiwengoh commented on cleanlab/classification.py:
```diff
@@ -610,7 +610,17 @@ def predict(self, *args, **kwargs) -> np.ndarray:
+            elif "X" in kwargs:
+                X = kwargs["X"]
+                del kwargs["X"]
+            elif "x" in kwargs:
+                X = kwargs["x"]
+                del kwargs["x"]
+            else:
+                raise ValueError("No input provided to predict")
```
I think we can stick with supporting the just same argname as sklearn ([example](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict)), so just checking for `X` should be fine, with a more descriptive error message otherwise.

huiwengoh replied:
```suggestion
            else:
                raise ValueError("No input provided to predict, please provide X.")
```

huiwengoh commented on cleanlab/classification.py:
```diff
@@ -631,7 +641,17 @@ def predict_proba(self, *args, **kwargs) -> np.ndarray:
+            elif "X" in kwargs:
+                X = kwargs["X"]
+                del kwargs["X"]
+            elif "x" in kwargs:
+                X = kwargs["x"]
+                del kwargs["x"]
+            else:
+                raise ValueError("No input provided to predict")
```
Same as above: 

```suggestion
            else:
                raise ValueError("No input provided to predict, please provide X.")
```

huiwengoh commented on tests/test_classification.py:
```diff
@@ -170,13 +170,32 @@ def test_cl(data):
+    result = cl.predict(X=HIGH_DIM_DATA["X_test"])
+    assert len(result) == len(HIGH_DIM_DATA["X_test"])
+
+    result = cl.predict(x=HIGH_DIM_DATA["X_test"])
+    assert len(result) == len(HIGH_DIM_DATA["X_test"])
```
Can remove since we will not support that

huiwengoh commented on tests/test_classification.py:
```diff
@@ -170,13 +170,32 @@ def test_cl(data):
+    pred_proba = cl.predict_proba(X=HIGH_DIM_DATA["X_test"])
+    assert pred_proba.shape == (len(HIGH_DIM_DATA["X_test"]), 10)
+
+    pred_proba = cl.predict_proba(x=HIGH_DIM_DATA["X_test"])
+    assert pred_proba.shape == (len(HIGH_DIM_DATA["X_test"]), 10)
```
Can remove since we will not support that:

```suggestion
```

huiwengoh commented:
Small change requested re the args supported, I think it is sufficient to support `X` like sklearn does (also our docstring specifies that we expect `X`).

After this change I think this PR is good to go! Great work @Steven-Yiran navigating some unexpected nuances!

Steven-Yiran commented:
@huiwengoh I just committed all the changes. Thanks again for your helpful comments.

huiwengoh reviewed (approved):
Added a small edit to make the `cl._default_clf` attribute private (personal preference + I think it makes it clearer that it is for internal use, ie. users don't have to worry about it), otherwise this PR looks great to me! 

Thanks for this improvement to CleanLearning @Steven-Yiran !
